### PR TITLE
Correctly generate PACKAGES information for R hosted repositories

### LINKED ISSUE #12: Package file generation does not give the latest version
Hi

So I have three packages in my hosted repo (at the moment, Jenkins pumps out a new one on each build) as follows:

image

curl -O http://**myip**:**myport**/repository/**myreponame**/bin/windows/contrib/3.4/PACKAGES.gz
#
the PACKAGES.gz file always shows the lowest version of the package not the highest:

```
Package: rClr
Version: 0.7-2
Depends: R (>= 3.1.0), methods
Suggests: testthat, knitr, xtable
License: LGPL-3

Package: rYieldLib
Version: 1.0.2017.154
Depends: R (>= 3.0.0), rClr, zoo, lubridate
Suggests: testthat, knitr, rmarkdown
NeedsCompilation: no

```

should be 

Package: rYieldLib
Version: 1.0.2017.15**5**

### PR DESCRIPTION
**Pull request by fjmilens3**
Initial attempt to implement correct generation of PACKAGES information for hosted repositories. 

This pull request makes the following changes:
* Adds an RPackageVersion class and associated tests for version comparisons.
* Adds an RPackagesBuilder to do the appropriate building of PACKAGES information for hosted.
* Uses the new RPackagesBuilder in RHostedFacetImpl when PACKAGES.gz requests are serviced.

Note that this could more immediately surface performance issues having to do with regenerating the PACKAGES.gz file on each request, as we have to maintain greatest-version information for each package in memory. My opinion would be that if we like this approach, we should merge it in, but also ensure that someone _volunteers_ to do the work for making the PACKAGES file generation work more like metadata for Yum Hosted.

Note that this seems to pass with the existing tests, but we should test it using actual R packages and the R client before we deem it correct (since we cannot run ITs for our open-source projects at this time). I would test manually (and will do if I have time) but I wanted to at least not let this sit here broken.

It relates to the following issue #s:
* Fixes #12

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -48,6 +48,11 @@
     </dependency>
 
     <dependency>
+      <groupId>se.sawano.java</groupId>
```
Comment (fjmilens3) [inline pom.xml]: Same dependency as we have for npm for doing the same basic comparison, which should (to my knowledge) also be suitable for this scenario (or at least no worse than it is for npm).

File: src/main/java/org/sonatype/nexus/repository/r/internal/RHostedFacetImpl.java [hunk 2]
```diff
@@ -75,35 +73,35 @@
   public Content getPackages(final String packagesPath) {
     checkNotNull(packagesPath);
     try {
+      // TODO: Do NOT do this on each request as there is at least some overhead, and memory usage is proportional to
```
Comment (fjmilens3) [inline src/main/java/org/sonatype/nexus/repository/r/internal/RHostedFacetImpl.java]: We need to do this, but probably deserves a separate PR.

File: src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilder.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  public void append(final Asset asset) {
+    // is this asset at this particular path?
+    String base = basePath(packagesPath);
+    if (base.equals(basePath(asset.name()))) {
+
+      // is this a newer (or equal) version of this asset's package than the one we currently have?
+      String packageName = asset.formatAttributes().get(P_PACKAGE, String.class);
+      RPackageVersion newVersion = new RPackageVersion(asset.formatAttributes().get(P_VERSION, String.class));
+      RPackageVersion oldVersion = packageVersions.getOrDefault(packageName, newVersion);
+      if (newVersion.compareTo(oldVersion) >= 0) {
```
Comment (doddi) [inline src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilder.java]: There is probably no need to check for >=, just > 0 should suffice

Comment (fjmilens3) [reply src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilder.java]: There's the scenario where you don't have an initial occurrence of _any_ version for that package, so this was my way of getting around it _sans_ a separate existence or null check.

(That said, if you didn't notice that in reviewing it, it may also be that this is too cryptic a means of implementing that, so maybe we do want an explicit check. I kind of like this in that it deals with that scenario and also keeps down the total number of conditionals.)

Comment (fjmilens3) [reply src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilder.java]: How do you feel about https://github.com/sonatype/nexus-repository-r/pull/18/commits/e0a43d9b8932856a24c07e70db0f259e2841c735 as an alternative? (Reading it now, I think this is probably the clearer one.)

File: src/main/java/org/sonatype/nexus/repository/r/internal/RHostedFacetImpl.java [hunk 4]
```diff
@@ -75,35 +73,35 @@
...
+      // the number of packages contained in a particular path. We should be able to generate this when something has
+      // changed or via a scheduled task using an approach similar to the Yum implementation rather than this method.
```
Comment (jlstephens89) [inline src/main/java/org/sonatype/nexus/repository/r/internal/RHostedFacetImpl.java]: I think we do both. Taking the implementation from Yum saves us having to do any extra setup to ensure the metadata is generated but having the scheduled task is a useful fall-back if for some reason the metadata gets corrupted or Nexus gets interrupted while it is generating the metadata.

File: src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilder.java (hunk 3 above)
Comment (doddi) [reply src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilder.java]: Much clearer and thanks for the original explanation
